src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,129 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    public static final int INITIAL_ARRAY_SIZE = 10;
+    public static final int RESIZE_NUMBER = 5;
+
+    private T[] arr;
+    private int elementsCount;
+
+    public ArrayList() {
+        arr = (T[]) new Object[INITIAL_ARRAY_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        checkIfNeedArrayResize();
+        arr[elementsCount++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexWhileAddElement(index);
+        checkIfNeedArrayResize();
+        shiftArrayToRight(index);
+        arr[index] = value;
+        elementsCount++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIfIndexCorrect(index);
+        return arr[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIfIndexCorrect(index);
+        arr[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexCorrect(index);
+        final T removedElement = arr[index];
+        shiftArrayToLeft(index);
+        arr[elementsCount - 1] = null;
+        elementsCount--;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        if (element == null) {
+            for (int i = 0; i < elementsCount; i++) {
+                if (arr[i] == null) {
```

Comment:
move logic to check inside findCorrectIndex() method

you can use || to check for any of multiple conditions